File: stable/backend/go/main_2.go
// Save this as main_2.go
package main

import (
    "database/sql"
    "os"
    "github.com/gofiber/fiber/v2"
    _ "github.com/mattn/go-sqlite3"
)

// Models
type Book struct {
    ID              int      `json:"id"`
    Title           string   `json:"title"`
    Author          string   `json:"author"`
    Category        string   `json:"category"`
    Description     string   `json:"description"`
    CoverImage      string   `json:"cover_image"`
    PublicationYear int      `json:"publication_year"`
    ISBN            string   `json:"isbn"`
    PDFPath         string   `json:"pdf_path"`
    Reviews         []Review `json:"reviews"`
}

type Review struct {
    ID     int    `json:"id"`
    BookID int    `json:"book_id"`
    Text   string `json:"text"`
    Author string `json:"author"`
}

// Database connection
var db *sql.DB

// Database Operations
func initDB() error {
    var err error
    db, err = sql.Open("sqlite3", "./books.db")
    if err != nil {
        return err
    }

    return createTables()
}

func closeDB() {
    if db != nil {
        db.Close()
    }
}

func createTables() error {
    createTablesSQL := `
    CREATE TABLE IF NOT EXISTS books (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        title TEXT NOT NULL,
        author TEXT NOT NULL,
        category TEXT NOT NULL,
        description TEXT,
        cover_image TEXT,
        publication_year INTEGER,
        isbn TEXT,
        pdf_path TEXT
    );

    CREATE TABLE IF NOT EXISTS reviews (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        book_id INTEGER,
        text TEXT NOT NULL,
        author TEXT NOT NULL,
        FOREIGN KEY(book_id) REFERENCES books(id)
    );`

    _, err := db.Exec(createTablesSQL)
    return err
}

func insertSampleData() error {
    var count int
    if err := db.QueryRow("SELECT COUNT(*) FROM books").Scan(&count); err != nil {
        return err
    }
    
    if count > 0 {
        return nil // Data already exists
    }

    tx, err := db.Begin()
    if err != nil {
        return err
    }
    defer tx.Rollback()

    bookStmt := `
    INSERT INTO books (
        title, author, category, description, cover_image, 
        publication_year, isbn, pdf_path
    ) VALUES (
        'The Sealed Nectar',
        'Safiur Rahman Mubarakpuri',
        'Biography',
        'Biography of Prophet Muhammad (peace be upon him)',
        'https://example.com/cover1.jpg',
        1979,
        '978-9960-899-55-8',
        './pdfs/sealed-nectar.pdf'
    )`

    result, err := tx.Exec(bookStmt)
    if err != nil {
        return err
    }

    bookID, err := result.LastInsertId()
    if err != nil {
        return err
    }

    reviewStmt := `
    INSERT INTO reviews (book_id, text, author) VALUES 
    (?, 'An excellent biography that provides detailed insights.', 'Abdullah')`

    if _, err := tx.Exec(reviewStmt, bookID); err != nil {
        return err
    }

    return tx.Commit()
}

func getBooks(c *fiber.Ctx) error {
    rows, err := db.Query(`
        SELECT id, title, author, category, description, 
               cover_image, publication_year, isbn, pdf_path 
        FROM books
    `)
    if err != nil {
        return fiber.NewError(fiber.StatusInternalServerError, "Failed to fetch books")
    }
    defer rows.Close()

    var books []Book
    for rows.Next() {
        var book Book
        if err := rows.Scan(
            &book.ID, &book.Title, &book.Author, &book.Category,
            &book.Description, &book.CoverImage, &book.PublicationYear,
            &book.ISBN, &book.PDFPath,
        ); err != nil {
            continue
        }
        
        if reviews, err := getBookReviews(book.ID); err == nil {
            book.Reviews = reviews
        }
        
        books = append(books, book)
    }

    if len(books) == 0 {
        books = []Book{}
    }

    return c.JSON(books)
}

func getBookReviews(bookID int) ([]Review, error) {
    rows, err := db.Query("SELECT id, text, author FROM reviews WHERE book_id = ?", bookID)
    if err != nil {
        return nil, err
    }
    defer rows.Close()

    var reviews []Review
    for rows.Next() {
        var review Review
        if err := rows.Scan(&review.ID, &review.Text, &review.Author); err != nil {
            return nil, err
        }
        review.BookID = bookID
        reviews = append(reviews, review)
    }

    return reviews, nil
}

func downloadPDF(c *fiber.Ctx) error {
    id := c.Params("id")
    
    var pdfPath string
    err := db.QueryRow("SELECT pdf_path FROM books WHERE id = ?", id).Scan(&pdfPath)
    if err != nil {
        return fiber.NewError(fiber.StatusNotFound, "Book not found")
    }

    if _, err := os.Stat(pdfPath); os.IsNotExist(err) {
        return fiber.NewError(fiber.StatusNotFound, "PDF file not found")
    }

    return c.Download(pdfPath)
}

func setupRoutes(app *fiber.App) {
    api := app.Group("/api/v1")
    
    books := api.Group("/books")
    books.Get("/", getBooks)
    books.Get("/:id/download", downloadPDF)
}

File: beta/backend/flask/main_2.py
# main_2.py
import os
import logging
import traceback
import sqlite3
from typing import List, Optional

# Configure logging
logging.basicConfig(
    level=logging.DEBUG,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    filename='app_logs/database_errors.log',
    filemode='a'
)
logger = logging.getLogger(__name__)

# Console handler for immediate visibility
console_handler = logging.StreamHandler()
console_handler.setLevel(logging.INFO)
formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
console_handler.setFormatter(formatter)
logger.addHandler(console_handler)

class Book:
    def __init__(self, row):
        self.id = row['id']
        self.title = row['title']
        self.author = row['author']
        self.category = row['category']
        self.description = row['description']
        self.cover_image = row['cover_image']
        self.publication_year = row['publication_year']
        self.isbn = row['isbn']
        self.pdf_path = row['pdf_path']
        self.reviews = []

    def to_dict(self):
        return {
            'id': self.id,
            'title': self.title,
            'author': self.author,
            'category': self.category,
            'description': self.description,
            'cover_image': self.cover_image,
            'publication_year': self.publication_year,
            'isbn': self.isbn,
            'reviews': [review.to_dict() for review in self.reviews]
        }

class Review:
    def __init__(self, row):
        self.id = row['id']
        self.book_id = row['book_id']
        self.text = row['text']
        self.author = row['author']

    def to_dict(self):
        return {
            'id': self.id,
            'book_id': self.book_id,
            'text': self.text,
            'author': self.author
        }

class DatabaseManager:
    @staticmethod
    def get_db_connection():
        try:
            conn = sqlite3.connect('books.db')
            conn.row_factory = sqlite3.Row
            logger.debug("Database connection established")
            return conn
        except sqlite3.Error as e:
            logger.error(f"Database connection error: {str(e)}")
            logger.error(traceback.format_exc())
            raise

    @staticmethod
    def init_db():
        try:
            logger.info("Initializing database")
            conn = DatabaseManager.get_db_connection()
            cursor = conn.cursor()
            
            # Create Books Table
            cursor.execute('''
            CREATE TABLE IF NOT EXISTS books (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                title TEXT NOT NULL,
                author TEXT NOT NULL,
                category TEXT NOT NULL,
                description TEXT,
                cover_image TEXT,
                publication_year INTEGER,
                isbn TEXT,
                pdf_path TEXT
            )''')

            # Create Reviews Table
            cursor.execute('''
            CREATE TABLE IF NOT EXISTS reviews (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                book_id INTEGER,
                text TEXT NOT NULL,
                author TEXT NOT NULL,
                FOREIGN KEY(book_id) REFERENCES books(id)
            )''')
            
            conn.commit()
            conn.close()
            logger.info("Database initialization complete")
        except Exception as e:
            logger.error(f"Database initialization error: {str(e)}")
            logger.error(traceback.format_exc())
            raise

    @staticmethod
    def insert_sample_data():
        try:
            logger.info("Inserting sample data")
            conn = DatabaseManager.get_db_connection()
            cursor = conn.cursor()

            # Check if books exist
            cursor.execute("SELECT COUNT(*) FROM books")
            if cursor.fetchone()[0] > 0:
                logger.info("Sample data already exists")
                conn.close()
                return

            # Sample books with varied details
            books_data = [
                {
                    'title': 'The Sealed Nectar',
                    'author': 'Safiur Rahman Mubarakpuri',
                    'category': 'Biography',
                    'description': 'Comprehensive biography of Prophet Muhammad (peace be upon him)',
                    'cover_image': 'https://example.com/sealed-nectar.jpg',
                    'publication_year': 1979,
                    'isbn': '978-9960-899-55-8',
                    'pdf_path': './pdfs/sealed-nectar.pdf'
                },
                {
                    'title': 'Clean Code',
                    'author': 'Robert C. Martin',
                    'category': 'Programming',
                    'description': 'A handbook of agile software craftsmanship',
                    'cover_image': 'https://example.com/clean-code.jpg',
                    'publication_year': 2008,
                    'isbn': '978-0132350884',
                    'pdf_path': './pdfs/clean-code.pdf'
                }
            ]

            for book in books_data:
                cursor.execute('''
                INSERT INTO books (
                    title, author, category, description, cover_image, 
                    publication_year, isbn, pdf_path
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                ''', tuple(book.values()))
                
                book_id = cursor.lastrowid

                # Insert sample reviews
                cursor.execute('''
                INSERT INTO reviews (book_id, text, author) 
                VALUES (?, ?, ?)
                ''', (book_id, f'Great book about {book["title"]}', 'Anonymous'))

            conn.commit()
            conn.close()
            logger.info("Sample data insertion complete")
        except Exception as e:
            logger.error(f"Sample data insertion error: {str(e)}")
            logger.error(traceback.format_exc())
            raise

    @staticmethod
    def get_book_reviews(book_id: int) -> List[Review]:
        try:
            logger.info(f"Retrieving reviews for book {book_id}")
            conn = DatabaseManager.get_db_connection()
            cursor = conn.cursor()
            
            cursor.execute("SELECT * FROM reviews WHERE book_id = ?", (book_id,))
            reviews = [Review(row) for row in cursor.fetchall()]
            
            conn.close()
            logger.info(f"Retrieved {len(reviews)} reviews for book {book_id}")
            return reviews
        except Exception as e:
            logger.error(f"Error retrieving reviews for book {book_id}: {str(e)}")
            logger.error(traceback.format_exc())
            return []

    @staticmethod
    def get_all_books() -> List[Book]:
        try:
            logger.info("Retrieving all books")
            conn = DatabaseManager.get_db_connection()
            cursor = conn.cursor()
            
            cursor.execute("SELECT * FROM books")
            books = [Book(row) for row in cursor.fetchall()]
            
            for book in books:
                book.reviews = DatabaseManager.get_book_reviews(book.id)
            
            conn.close()
            logger.info(f"Retrieved {len(books)} books")
            return books
        except Exception as e:
            logger.error(f"Error retrieving all books: {str(e)}")
            logger.error(traceback.format_exc())
            return []

    @staticmethod
    def get_book_by_id(book_id: int) -> Optional[Book]:
        try:
            logger.info(f"Attempting to retrieve book with ID: {book_id}")
            conn = DatabaseManager.get_db_connection()
            cursor = conn.cursor()
            
            cursor.execute("SELECT * FROM books WHERE id = ?", (book_id,))
            result = cursor.fetchone()
            
            if not result:
                logger.warning(f"No book found with ID: {book_id}")
                return None
            
            book = Book(result)
            book.reviews = DatabaseManager.get_book_reviews(book.id)
            logger.info(f"Book retrieved: {book.title} (ID: {book.id})")
            return book
        except Exception as e:
            logger.error(f"Error retrieving book by ID {book_id}: {str(e)}")
            logger.error(traceback.format_exc())
            return None

    @staticmethod
    def delete_book(book_id: int) -> bool:
        try:
            logger.info(f"Attempting to delete book with ID: {book_id}")
            conn = DatabaseManager.get_db_connection()
            cursor = conn.cursor()
            
            cursor.execute("DELETE FROM books WHERE id = ?", (book_id,))
            conn.commit()
            conn.close()
            logger.info(f"Book with ID {book_id} deleted successfully")
            return True
        except Exception as e:
            logger.error(f"Error deleting book with ID {book_id}: {str(e)}")
            logger.error(traceback.format_exc())
            return False

    @staticmethod
    def update_book(book_id: int, updated_data: dict) -> bool:
        try:
            logger.info(f"Attempting to update book with ID: {book_id}")
            conn = DatabaseManager.get_db_connection()
            cursor = conn.cursor()
            
            cursor.execute('''
            UPDATE books SET title = ?, author = ?, category = ?, 
            description = ?, cover_image = ?, publication_year = ?, 
            isbn = ?, pdf_path = ? WHERE id = ?
            ''', (*updated_data.values(), book_id))
            
            conn.commit()
            conn.close()
            logger.info(f"Book with ID {book_id} updated successfully")
            return True
        except Exception as e:
            logger.error(f"Error updating book with ID {book_id}: {str(e)}")
            logger.error(traceback.format_exc())
            return False
